File: matlab/problem1/modelFitVal.m
function pCorrect = modelFitVal(X, L, k)

% modelFitVal takes multivariate data (one feature per column) and
% associated labels and returns the cross-validated performance of a linear
% model.

% create a random permutation of sample indices
nSamples = size(X, 1);
randIdx = ...

% initialize performance measure
pCorrect = 0;

for iPart=1:k
    % construct 2 vectors of indices that divide the samples into
    % validation and training set of size nSamples/k and (k-1)*nSamples/k,
    % respectively. for uneven divisions, make sure set sizes never differ
    % by more than 1 over all iterations
    idxTest = ...
    idxTrain = ...
    % select data and labels using the above index vectors
    xTest = X(...);
    xTrain = X(...);
    lTest = L(...);
    lTrain = L(...);
    % fit logit model to training data
    coeff = glmfit(...);
    % compute labels for test data from rounded p(C1|D)
    lPredicted = round(...);
    % compare predicted to actual labels and accumulate performance
    pCorrect = pCorrect + 1/k*mean(lPredicted==lTest);
end

File: python/problem1/modelFitVal.py
from pylab import *
from numpy.random import shuffle
import statsmodels.api as sm


def modelFitVal(X, L, kPart):
    """
    modelFitVal takes multivariate data (one feature per column) and
    associated labels and returns the cross-validated performance of a linear
    model.

    """

    # set up shuffled data partitions
    nSamples = ...
    partTag = ...
    shuffle(partTag)

    pCorrect = 0

    for iPart in range(kPart):

        # separate training and test data
        index = ...
        xTest = X[...]
        lTest = L[...]
        xTrain = X[~...]
        lTrain = L[~...]

        # add columns of ones
        xTest = sm.add_constant(xTest)
        xTrain = sm.add_constant(xTrain)

        # fit model to training data using logistic regression
        model = sm.GLM(...)
        model = ...

        # compute p(C1|D) for test data
        posterior_C1 = model.predict(xTest)

        # compare model's classification to actual labels
        pCorrect += mean(posterior_C1.round() == lTest) / kPart

    return pCorrect
